lib/os/queued_operation.c

```diff
@@ -0,0 +1,197 @@
+		 */
+		if (mgr->finalizing) {
+			break;
+		}
+
+		sys_snode_t *node = sys_slist_get(&mgr->operations);
+
+		if (node) {
+			op = CONTAINER_OF(node, struct queued_operation, node);
+			mgr->current = op;
+		} else {
+			op = NULL;
+		}
+
+		k_spin_unlock(&mgr->lock, key);
+
+		/* Only notify the manager if there's an operation, or
+		 * if it is to transition to idle.
+		 */
+		if ((op != NULL) || !in_idle) {
+			mgr->vtable->process(mgr, op);
+		}
+		in_idle = (op == NULL);
+
+		key = k_spin_lock(&mgr->lock);
+	} while (op != NULL);
+
+	k_spin_unlock(&mgr->lock, key);
+}
+
+int queued_operation_submit(struct queued_operation_manager *mgr,
+			    struct queued_operation *op,
+			    int priority)
+{
+	int rv = 0;
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(mgr->vtable != NULL);
+	__ASSERT_NO_MSG(mgr->vtable->process != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	/* Validation is optional; if present, use it. */
+	if (mgr->vtable->validate) {
+		rv = mgr->vtable->validate(mgr, op);
+	}
+
+	/* Set the priority, checking whether it's in range. */
+	if (rv == 0) {
+		rv = op_set_priority(op, priority);
+	}
+
+	/* Reject callback notifications without translation
+	 * function.
+	 */
+	if ((rv == 0)
+	    && async_notify_uses_callback(&op->notify)
+	    && (mgr->vtable->callback == NULL)) {
+		rv = -ENOTSUP;
+	}
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+	sys_slist_t *list = &mgr->operations;
+	struct queued_operation *prev = NULL;
+	struct queued_operation *tmp;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(list, tmp, node) {
+		if (priority < op_get_priority(tmp)) {
+			break;
+		}
+		prev = tmp;
+	}
+
+	if (prev == NULL) {
+		sys_slist_prepend(list, &op->node);
+	} else {
+		sys_slist_insert(list, &prev->node, &op->node);
+	}
+
+	select_next_and_unlock(mgr, key);
+
+	return rv;
+}
+
+void queued_operation_finalize(struct queued_operation_manager *mgr,
+			       struct queued_operation *op,
+			       int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	if (mgr->current == op) {
+		mgr->finalizing = true;
+		mgr->current = NULL;
+	}
+
+	k_spin_unlock(&mgr->lock, key);
+
+	async_notify_generic_callback cb = async_notify_finalize(&op->notify, res);
+
+	if (cb != NULL) {
+		mgr->vtable->callback(mgr, op, cb);
+	}
+
+	key = k_spin_lock(&mgr->lock);
+
+	mgr->finalizing = false;
+
+	if (can_start(mgr)) {
+		select_next_and_unlock(mgr, key);
+	} else {
+		k_spin_unlock(&mgr->lock, key);
+	}
+}
+
+int queued_operation_cancel(struct queued_operation_manager *mgr,
+			    struct queued_operation *op)
+{
+	int rv = 0;
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	if (op == mgr->current) {
+		rv = -EINPROGRESS;
+	} else if (!sys_slist_find_and_remove(&mgr->operations, &op->node)) {
+		rv = -EINVAL;
+	}
+
+	k_spin_unlock(&mgr->lock, key);
+
+	if (rv == 0) {
+		queued_operation_finalize(mgr, op, -ECANCELED);
```

Comment:
why not just call 
```
async_notify_generic_callback cb = async_notify_finalize(&op->notify, res);

if (cb != NULL) {
	mgr->vtable->callback(mgr, op, cb);
}
```

now, if it interrupts `queued_operation_finalize` called from other context which will set finalize to true then this will clear it prematurely.

Reply:
Good catch; fixed.

Thanks.  All documentation comments in this review have been addressed.  Please resolve what you think is done, and provide more information if you think there's something that wasn't addressed.